src/IO/CachedInMemoryReadBufferFromFile.cpp

```diff
@@ -0,0 +1,186 @@
+#include "CachedInMemoryReadBufferFromFile.h"
+#include <IO/SwapHelper.h>
+#include <base/scope_guard.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNEXPECTED_END_OF_FILE;
+    extern const int CANNOT_SEEK_THROUGH_FILE;
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+}
+
+CachedInMemoryReadBufferFromFile::CachedInMemoryReadBufferFromFile(
+    FileChunkAddress cache_key_, PageCachePtr cache_, std::unique_ptr<ReadBufferFromFileBase> in_, const ReadSettings & settings_)
+    : ReadBufferFromFileBase(0, nullptr, 0, in_->getFileSize()), cache_key(cache_key_), cache(cache_), settings(settings_), in(std::move(in_))
+    , read_until_position(file_size.value())
+{
+    chassert(file_size.has_value());
```

Comment:
Looks like this assertion will never fail as above `read_until_position(file_size.value())` will fail first.